src/analyses/expsplit.ml

```diff
@@ -0,0 +1,86 @@
+open Prelude.Ana
+open Analyses
+
+module M = Messages
+
+module ID = Queries.ID
+
+module Spec : Analyses.MCPSpec =
+struct
+  let name () = "expsplit"
+
+  module D = MapDomain.MapBot (Basetype.CilExp) (ID)
+  module C = D
+
+  let startstate v = D.bot ()
+  let exitstate = startstate
+
+  include Analyses.DefaultSpec
+
+  let should_join = D.equal
+
+  let emit_splits ctx d =
+    D.iter (fun e _ ->
+        ctx.emit (UpdateExpSplit e)
+      ) d;
+    d
+
+  let emit_splits_ctx ctx =
+    emit_splits ctx ctx.local
+
+  let assign ctx (lval:lval) (rval:exp) =
+    emit_splits_ctx ctx
+
+  let vdecl ctx (var:varinfo) =
+    emit_splits_ctx ctx
+
+  let vdecl ctx (var:varinfo) =
+    emit_splits_ctx ctx
+
+  let branch ctx (exp:exp) (tv:bool) =
+    emit_splits_ctx ctx
+
+  let enter ctx (lval: lval option) (f:fundec) (args:exp list) =
+    [ctx.local, ctx.local]
+
+  let body ctx (f:fundec) =
+    emit_splits_ctx ctx
+
+  let return ctx (exp:exp option) (f:fundec) =
+    emit_splits_ctx ctx
+
+  let combine ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc au =
+    let d = D.join ctx.local au in
+    emit_splits ctx d
```

Comment:
I'm not sure how the interprocedurality of this should be. Right now the splitting expressions are passed to functions and are in contexts. They're also joined here on `combine`, in case the function added more splits. But this is practically quite limited because it only works on globals, because locals go out of scope and thus evaluate the same either way.

It might be more interesting with var_eq information to modify the splitting accordingly when going over function boundaries, but that can be a future experiment.